spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -102,12 +115,39 @@ public void onMessage(org.springframework.amqp.core.Message amqpMessage, Channel
 		if (logger.isDebugEnabled()) {
 			logger.debug("Processing [" + message + "]");
 		}
-		Object result = invokeHandler(amqpMessage, channel, message);
-		if (result != null) {
-			handleResult(result, amqpMessage, channel, message);
+		try {
+			Object result = invokeHandler(amqpMessage, channel, message);
+			if (result != null) {
+				handleResult(result, amqpMessage, channel, message);
+			}
+			else {
+				logger.trace("No result object given - no result to handle");
+			}
 		}
-		else {
-			logger.trace("No result object given - no result to handle");
+		catch (ListenerExecutionFailedException e) {
+			if (this.errorHandler != null) {
+				try {
+					Object result = this.errorHandler.handleError(amqpMessage, message, e);
+					if (result != null) {
+						handleResult(result, amqpMessage, channel, message);
+					}
+					else {
+						logger.trace("Error handler returned no result");
+					}
+				}
+				catch (Exception ex) {
+					if (!this.propagateExceptions) {
```

Comment:
Might be rename `propagateExceptions` everywhere to something more explicit, e.g. `returnExceptionsToSender`?

To me _if not propagate then throw_ looks confusing, as quite often it's written the other way around.
After rename it can be written as _if returnExceptionsToSender then handleResult(...) else throw_ and it's clear what it does.